### ISSUE TITLE
Adding new node to the running cluster

### ISSUE THREAD

**Issue by Amandine-G** [arangodb/arangodb#4876]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.3

Mode:
- [x] Cluster

Storage-Engine:
- [x] rocksdb

On this operating system:

  - [x] other: centos 6.5

We now have a cluster running with 3 nodes, and we want to add another 2 new nodes to the cluster. I tried to start the node by  `arangodb --starter.join 65.18.27.28:8529`  while the cluster is running 
but failed.
the log printed were as follows:

> 2018/03/16 17:09:57 Starting arangodb version 0.10.4, build 553aab6
2018/03/16 17:09:57 Contacting master http://65.18.27.28:8528...
2018/03/16 17:09:57 Cannot start because of error from master: Post http://65.18.27.28:8528/hello: dial tcp 65.18.27.28:8528: getsockopt: connection refused

In case the master node is one of the other working nodes, I tried with all the other 2 ip addresses,but got the same result.
Should I stop the cluster before adding new nodes?
I didn't found much info  on line, somebody help, please...

**Comment by OmarAyo** [comment]
Hi @Amandine-G 

Thank you for using ArangoDB

In the command that you are using to add another node you are missing _--starter.data-dir=path to your node_ 

So use the following command to add more nodes

```
arangodb. --starter.data-dir=path to your node --starter.join 127.0.0.1

```

Best

**Comment by Amandine-G** [comment]
Hi @OmarAyo,

Thank you for your response!
I was running the command in the data path,so I didn't use `--starter.data-dir` in the command at first. But I tried with your suggestion, sadly still got the same response.

I didn't find any log mentioning the new node. So what else should I check or modify?

**Comment by sleto-it** [comment]
Hi,

To reply one of your question: to add a new node it is not needed to stop the cluster

I wonder if you can share with us the commands that you have used to start the other nodes?

I assume you are starting this new node from a different machine, is that correct? Are you able to ping from this (new) machine the others?

Are you using JWT auth?

Thanks,

**Comment by Amandine-G** [comment]
Hi,
I added 2 nodes to my cluster successfully!
First, I excuted the command `arangodb --starter.address 65.18.27.28 --starter.data-dir=/arangodb_data` at my assuming master node.

> 2018/03/17 16:44:24 Starting arangodb version 0.10.4, build 553aab6
> 2018/03/17 16:44:24 Relaunching service with id 'bdc504a8' on 65.18.27.28:8528...
> 2018/03/17 16:44:24 Looking for a running instance of agent on port 8531
> 2018/03/17 16:44:24 Listening on 0.0.0.0:8528 (65.18.27.28:8528)
> 2018/03/17 16:44:24 agent seems to be running already, checking port 8531...
> 2018/03/17 16:44:24 agent is already running on 8531. No need to start anything.
> 2018/03/17 16:44:24 agent up and running (version 3.3.1).
> 2018/03/17 16:44:25 Just became master
> 2018/03/17 16:44:25 Looking for a running instance of dbserver on port 8530
> 2018/03/17 16:44:25 dbserver seems to be running already, checking port 8530...
> 2018/03/17 16:44:25 dbserver is already running on 8530. No need to start anything.
> 2018/03/17 16:44:25 dbserver up and running (version 3.3.1).
> 2018/03/17 16:44:26 Looking for a running instance of coordinator on port 8529
> 2018/03/17 16:44:26 coordinator seems to be running already, checking port 8529...
> 2018/03/17 16:44:26 coordinator is already running on 8529. No need to start anything.
> 2018/03/17 16:44:26 coordinator up and running (version 3.3.1).
> 2018/03/17 16:44:26 Your cluster can now be accessed with a browser at `http://65.18.27.28:8529` or
> 2018/03/17 16:44:26 using `arangosh --server.endpoint tcp://65.18.27.28:8529`.

After it writed log like the above, I retried your suggested command `arangodb --starter.data-dir=/arangodb_data  --starter.join 65.18.27.28  --server.storage-engine=rocksdb `, and it printed the following log at the new node.

> 2018/03/17 16:44:35 Starting arangodb version 0.10.4, build 553aab6
> 2018/03/17 16:44:35 Contacting master http://65.18.27.28:8528...
> 2018/03/17 16:44:35 Waiting for 3 servers to show up...
> 2018/03/17 16:44:35 Serving as slave with ID '6eea5223' on :8528...
> 2018/03/17 16:44:35 Listening on 0.0.0.0:8528 (:8528)
> 2018/03/17 16:44:35 Looking for a running instance of dbserver on port 8530
> 2018/03/17 16:44:35 Starting dbserver on port 8530
> 2018/03/17 16:44:36 Looking for a running instance of coordinator on port 8529
> 2018/03/17 16:44:36 Starting coordinator on port 8529
> 2018/03/17 16:44:37 dbserver up and running (version 3.3.1).
> 2018/03/17 16:44:38 coordinator up and running (version 3.3.1).
> 2018/03/17 16:44:38 Your cluster can now be accessed with a browser at `http://65.18.27.27:8529` or
> 2018/03/17 16:44:38 using `arangosh --server.endpoint tcp://65.18.27.27:8529`.

and the master nodes shows

> 2018/03/17 16:44:35 Added new peer '6eea5223': 65.18.27.27, portOffset: 0

I don't quit understand why it has to run the first command at the master node. While excuting `arangodb`  on one already running server , what it's accually doing in the background ?

And for the question of @sleto-it, I am not using JWT auth. And yes, I'm starting this new node from a different machine, they ping well between each other.


________________________________
发件人: sleto-it <notifications@github.com>
发送时间: 2018年3月17日 16:20:06
收件人: arangodb/arangodb
抄送: Amandine-G; Mention
主题: Re: [arangodb/arangodb] Adding new node to the running cluster (#4876)


Hi,

To reply one of your question: to add a new node it is not needed to stop the cluster

I wonder if you can share with us the commands that you have used to start the other nodes?

I assume you are starting this new node from a different machine, is that correct? Are you able to ping from this (new) machine the others?

Are you using JWT auth?

Thanks,

―

**Comment by OmarAyo** [comment]
Hi @Amandine-G 

I am happy to know that you have been able to add 2 more nodes.
Regarding your aforementioned misunderstandings let me do some tests and get back to you as soon as possible.

In the meanwhile, feel free to ask any other related question to this ticket.

Best,
Omar

**Comment by Amandine-G** [comment]
Hi @Omar_A,

Thank you for your response.

I noticed that on the 2 newly added nodes, no agents were created. So in the cluster, there are  in total 3 agents, 5 Coordinators and 5 DB Servers.

If I didn't take it wrong, as there is not agent created, the 2 new nodes could not become a master, is that correct?

If I use `--cluster.start-agent true` to restart this 2 nodes will them function well? I would also test it.

**Comment by OmarAyo** [comment]
Hi @Amandine-G 

I will divide your response into two different parts and I will give you a response for them:

- **I noticed that on the 2 newly added nodes, no agents were created. So in the cluster, there are  in total 3 agents, 5 Coordinators and 5 DB Servers.**

Yes, with 5 nodes you have 3 Agents, 5 Coordinators and 5 DB servers. This is because by default the agency size is 3, if you want/need to increase your agency size you just have to add to your first command line 
```--cluster.agency-size X``` (for X I mean an integer, odd number) -  so in your case you have to add 
```--cluster.agency-size 5```

Please note that this command has to be run only at the first node.  

You can use ```--cluster.start-agent true``` as well to have agents on the newly nodes.


-  **If I didn't take it wrong, as there is not agent created, the 2 new nodes could not become a master, is that correct?**

Probably you are making some confusion here, please note that you can have neither _master_ nor _slave_ in ArangoDB Cluster deployment.    
When you read the message _Just became master_ from one of your nodes, this is referring to the _Starter_ and not to the ArangoDB Cluster, so you do not have to worry about who is the _master_ since you do not have a _master_ in the ArangoDB Cluster.

In conclusion, I think that 3 Agents are enough to handle 5 nodes so no more action is needed since you have an ArangoDB Cluster with 5 nodes up and running. 

For any additional question, feel free to ask.

Best.

**Comment by Amandine-G** [comment]
Hi @Omar_A,
Thank you for your explanation,that makes more sense.
I'd keep my cluster with 3 agengts.

**Comment by OmarAyo** [comment]
Hi @Amandine-G

We are here to help you ! 
So, I am going to close this ticket, if you have any other question feel free to reopen it at any time.

Best,
Omar

**Comment by OmarAyo** [comment]
Hi @Amandine-G 

Could you please describe in which node did you use the following command:  ``` --cluster.agency-size 5 ``` ?

**Comment by Amandine-G** [comment]
Hi @OmarAyo,
Thank you for your response!
Since we didn't found a way to control memory use, if one of my nodes reached 90G of memory consumption, We'd restart the node manually. For example, yesterday we used `top |grep arangod` to check my nodes A, B, C, D and E, we found that on node A, arangodb consumed 93G of memory. So we shutted node A down with `arangodb stop` , normally the cluster should work as usual, but yesterday the other nodes became really slow, the web UI can not refresh, and even excuting an easy AQL took long time. We assumed that the cluster was adjusting to cover the absent of node A. But after 30 mins, it was still the same. So we decided to restart the whole cluster. 
We tried several times to restart the cluster. At first it didn't work, so we wrote a question on StackOverFlow[(qustion),](https://stackoverflow.com/questions/49507915/arangodb-cluster-restart-failure) but at last it restarted. So right now, the cluster is up and running. But it's really confusing for us.
We are in the testing process of using arangodb, we want to make sure that if 3/5 nodes break down, the rest 2 nodes can work normally. i.e. if the 3 nodes with agents broke down, the cluster would not have agents, we want to make sure the cluster can handle this kind of circumstance.

**Comment by OmarAyo** [comment]
Hi @Amandine-G,

Thank you for the detailed response

Now I can understand why you are trying to increase the size of your agency in the ArangoDB Cluster

Please let's discuss about the memory issue in a separate ticket

If you want to be sure that your 5 nodes cluster is up an running when bringing down two (random) nodes, then an agency size = 5 is what you need

The Cluster cannot work if the Agency is not up and running. The Agency makes use of the RAFT protocol. If your Agency is made of 3 Agents, then if two are down then the Agency is down (and same for your Cluster). If you Agency is made of 5 Agents, then if two agents are down the Agency will survive (and same for your Cluster)

If you want to survive to 3 machines down, then other setups are possible

You can also consider using separate machines for the Agency, e.g.:

- 3 dedicated machines for the Agency 
- plus additional 3 machines for DBServers+Coordinators (total 6 machines) with Replication factor = 3

The above setup will survive to 1 Agent down and 2 DBServers down (so 3 machines down in total)

Please let me know if this helps and feel free to ask more

Best

**Comment by Amandine-G** [comment]
Hi @OmarAyo ，
Thank you for you explanation!
I'd close this issue, and open a new ticket for the memory problem.
